src/AggregateFunctions/AggregateFunctionTimeseriesInstantValue.h

```diff
@@ -0,0 +1,174 @@
+        if (time_difference == 0)
+        {
+            result = 0;
+            null = 1;
+            return;
+        }
+
+        /// Resets must be take into account for `irate` function because it expects counter timeseries that only increase.
+        /// But `idelta` function expects gauge timeseries that can decrease and it is not considered to be a reset.
+        constexpr bool adjust_to_resets = is_rate;
+
+        ValueType value_difference = (adjust_to_resets && value < previous_value) ? value : (value - previous_value);
+        result = value_difference;
+        if constexpr (is_rate)
+            result = result * Base::timestamp_scale_multiplier / time_difference;
+        null = 0;
+    }
+
+    /// Insert the result into the column
+    /// timestamps buffer is reused between calls to avoid unnecessary allocations/deallocations
+    void doInsertResultInto(AggregateDataPtr __restrict place, IColumn & to) const
+    {
+        ColumnArray & arr_to = typeid_cast<ColumnArray &>(to);
+        ColumnArray::Offsets & offsets_to = arr_to.getOffsets();
+
+        offsets_to.push_back(offsets_to.back() + Base::bucket_count);
+
+        if (!Base::bucket_count)
+            return;
+
+        ColumnNullable & result_to = typeid_cast<ColumnNullable &>(arr_to.getData());
+        auto & data_to = typeid_cast<typename Base::ColVecResultType &>(result_to.getNestedColumn()).getData();
+        auto & nulls_to = result_to.getNullMapData();
+
+        const size_t old_size = data_to.size();
+        chassert(old_size == nulls_to.size(), "Sizes of nested column and null map of Nullable column are not equal");
+
+        data_to.resize(old_size + Base::bucket_count);
+        nulls_to.resize(old_size + Base::bucket_count);
+
+        ValueType * values = data_to.data() + old_size;
+        UInt8 * nulls = nulls_to.data() + old_size;
+
+        const auto & buckets = Base::data(place)->buckets;
+
+        /// Fill the data for missing buckets
+        TimestampType current_timestamp = Base::start_timestamp;
+        Bucket last_2_samples; /// Sliding window with last 2 samples
+        for (size_t i = 0; i < Base::bucket_count; ++i, current_timestamp += Base::step)
+        {
+            values[i] = ValueType{};
+            nulls[i] = 1;
+
+            auto bucket_it = buckets.find(i);
+            if (bucket_it != buckets.end())
+                last_2_samples.merge(bucket_it->second);
+
+            /// If the oldest of last 2 samples is within the window, we can calculate the rate or delta
+            if (last_2_samples.filled == 2 && last_2_samples.timestamps[1] + Base::window >= current_timestamp)
+            {
+                fillResultValue(last_2_samples.timestamps[0], last_2_samples.values[0],
+                    last_2_samples.timestamps[1], last_2_samples.values[1],
+                    values[i], nulls[i]);
+            }
+        }
+    }
+
+    static constexpr UInt16 FORMAT_VERSION = 2;
```

Comment:
When `FORMAT_VERSION` was `1`?

Reply:
It was used during testing with an early adopter customer, so for compatibility it makes sense to not reset it.